updater/lib/dependabot/dependency_group_engine.rb

```diff
@@ -5,81 +5,80 @@
 # This class implements our strategy for keeping track of and matching dependency
 # groups that are defined by users in their dependabot config file.
 #
-# This is a static class tied to the lifecycle of a Job
-# - Each UpdateJob registers its own DependencyGroupEngine which calculates
-#    the grouped and ungrouped dependencies for a DependencySnapshot
-# - Groups are only calculated once after the Job has registered its dependencies
-# - All allowed dependencies should be passed in to the calculate_dependency_groups! method
+# We instantiate the DependencyGroupEngine after parsing dependencies, configuring
+# any groups from the job's configuration before assigning the dependency list to
+# the groups.
+#
+# We permit dependencies to be in more than one group and also track those which
+# have zero matches so they may be updated individuall.
 #
 # **Note:** This is currently an experimental feature which is not supported
 #           in the service or as an integration point.
 #
 module Dependabot
-  module DependencyGroupEngine
-    @groups_calculated = false
-    @registered_groups = []
-
-    @dependency_groups = {}
-    @ungrouped_dependencies = []
+  class DependencyGroupEngine
+    class ConfigurationError < StandardError; end
 
-    def self.reset!
-      @groups_calculated = false
-      @registered_groups = []
-
-      @dependency_groups = {}
-      @ungrouped_dependencies = []
-    end
+    def self.from_job_config(job:)
+      groups = job.dependency_groups.map do |group|
```

Comment:
Looks like the smoke tests are upset that `dependency_groups` is `nil`. Do we need to default that to `[]`?

Reply:
Fixed!